com.unity.shadergraph/Editor/Drawing/Inspector/PropertyDrawers/ShaderInputPropertyDrawer.cs

```diff
@@ -1450,10 +1457,203 @@ string GetSanitizedReferenceName(string name)
+                        Debug.LogWarning("Invalid display name. Display names cannot be empty or all whitespace.");
+                    else if (int.TryParse(displayName, out int intVal) || float.TryParse(displayName, out float floatVal))
+                        Debug.LogWarning("Invalid display name. Display names cannot be valid integer or floating point numbers.");
+                    else
+                        dropdown.entries[index] = new DropdownEntry(GetFirstUnusedDropdownID(), displayName);
+
+                    this._postChangeValueCallback(true);
+                }
+            };
+
+            // Element height
+            m_DropdownReorderableList.elementHeightCallback = (int indexer) =>
+            {
+                return m_DropdownReorderableList.elementHeight;
+            };
+
+            // Can add
+            m_DropdownReorderableList.onCanAddCallback = (ReorderableList list) =>
+            {
+                return true;
+            };
+
+            // Can remove
+            m_DropdownReorderableList.onCanRemoveCallback = (ReorderableList list) =>
+            {
+                return list.count > DropdownNode.k_MinEnumEntries;
+            };
+
+            // Add callback delegates
+            m_DropdownReorderableList.onSelectCallback += DropdownSelectEntry;
+            m_DropdownReorderableList.onAddCallback += DropdownAddEntry;
+            m_DropdownReorderableList.onRemoveCallback += DropdownRemoveEntry;
+            m_DropdownReorderableList.onReorderCallback += DropdownReorderEntries;
+        }
+
+        void DropdownSelectEntry(ReorderableList list)
+        {
+            m_DropdownSelectedIndex = list.index;
+        }
+
+        int GetFirstUnusedDropdownID()
+        {
+            if (!(shaderInput is ShaderDropdown dropdown))
+                return 0;
+
+            List<int> ids = new List<int>();
+
+            foreach (DropdownEntry dropdownEntry in dropdown.entries)
+            {
+                ids.Add(dropdownEntry.id);
+            }
+
+            for (int x = 1;; x++)
+            {
+                if (!ids.Contains(x))
+                    return x;
+            }
+        }
+
+        void DropdownAddEntry(ReorderableList list)
+        {
+            if (!(shaderInput is ShaderDropdown dropdown))
+                return;
+
+            this._preChangeValueCallback("Add Dropdown Entry");
+
+            int index = GetFirstUnusedDropdownID();
+
+            var displayName = GetDuplicateSafeDropdownDisplayName(index, "New");
+
+            // Add new entry
+            dropdown.entries.Add(new DropdownEntry(index, displayName));
+
+            // Update GUI
+            this._postChangeValueCallback(true);
+            this._dropdownChangedCallback();
+            m_DropdownSelectedIndex = list.list.Count - 1;
+        }
+
+        void DropdownRemoveEntry(ReorderableList list)
+        {
+            if (!(shaderInput is ShaderDropdown dropdown))
```

Comment:
Hmm in fact given that you check for whether the input is a dropdown at the entry point to all of this code (i.e. BuildDropdownFields) all these checks after that point might be redundant? unless they are also meant to function as null checks in case a dropdown is somehow invalidated suddenly

Reply:
Possibly, I just followed the pattern used by keywords, so I assumed it might be needed?

Reply:
Would prefer to leave this in place out of caution.